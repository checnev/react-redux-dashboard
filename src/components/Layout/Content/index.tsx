import React from 'react';
import {
  Container,
  Grid,
} from '@material-ui/core';

interface ContentProps {
  children: React.ReactNode;
  sidebar?: React.ReactNode;
}

const Content: React.FC<ContentProps> = ({ children, sidebar: SideBar}: ContentProps) => {
  const contentSize = SideBar? 9 : 12;

  return (
    <Container maxWidth="lg">
      <Grid container>
        { SideBar && <Grid item xs={12} sm={3}>{SideBar}</Grid> }
        <Grid item xs={12} sm={contentSize}>{children}</Grid>
      </Grid>
    </Container>
  );
};

export default Content;
