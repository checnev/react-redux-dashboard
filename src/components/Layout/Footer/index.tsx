import {
  Container,
  Typography,
  Divider,
} from '@material-ui/core';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

const useStyles = makeStyles((theme: Theme) => createStyles({
  root: {
    marginTop: 'auto',
  },
  container: {
    paddingTop: 20,
    paddingBottom: 20,
  },
  policy: {
    fontSize: '.9rem',
  },
}));

const Footer = () => {
  const classes = useStyles();
  const year = new Date().getFullYear();

  return (
    <footer className={classes.root}>
      <Divider />
      <Container maxWidth="lg" className={classes.container}>
        <Typography align="right">CarShare Â© {year}</Typography>
      </Container>
    </footer>
  );
};

export default Footer;
