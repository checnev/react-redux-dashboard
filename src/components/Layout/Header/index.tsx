import React from 'react';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import {
  AppBar,
  Toolbar,
  Typography,
  IconButton,
} from '@material-ui/core';
import {
  Menu as MenuIcon,
  Brightness4,
  Brightness7,
} from '@material-ui/icons';


const useStyles = makeStyles((theme: Theme) => createStyles({
  root: {
  },
  menuButton: {
    marginRight: theme.spacing(2),
  },
  title: {
    flexGrow: 1,
  },
  iconChangeTheme: {},
}));

interface HeaderProps {
  openDrawer?: () => void;
  changeTheme?: () => void;
  themeMode: 'dark' | 'light';
}

const Header: React.FC<HeaderProps> = React.memo(
  ({ openDrawer, changeTheme, themeMode }: HeaderProps) => {
  const classes = useStyles();

  return (
    <AppBar position="relative" className={classes.root}>
      <Toolbar>
        <IconButton
          edge="start"
          className={classes.menuButton}
          color="inherit"
          aria-label="menu"
          onClick={openDrawer}
        >
          <MenuIcon />
        </IconButton>
        <Typography variant="h6" className={classes.title}>
          CarShare
        </Typography>
        <IconButton
          aria-label="change theme"
          onClick={changeTheme}
          color="inherit"
        >
          {themeMode === 'light' ? <Brightness4 /> : <Brightness7 />}
        </IconButton>
      </Toolbar>
    </AppBar>
  );
}
);

export default Header;
