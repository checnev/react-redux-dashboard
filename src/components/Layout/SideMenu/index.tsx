import React from 'react';
import {
  Drawer,
  MenuList,
  MenuItem,
  ListItemIcon,
  Typography,
  Link
} from '@material-ui/core';
import { NavLink } from 'react-router-dom';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import { 
  Dashboard,
  ViewList,
  Report,
  Map,
} from '@material-ui/icons';

const useStyles = makeStyles((theme: Theme) => {
  return createStyles({
  drawer: {
    width: 240,
  },
  drawerPaper: {
    width: 240,
  },
  menuList: {
    paddingTop: 0,
    '&:active': {
      outline: 'none',
    },
  },
  menuItem: {
    padding: 0,
  },
  link: {
    display: 'flex',
    flexGrow: 1,
    alignItems: 'center',
    textDecoration: 'none',
    padding: '6px 16px',
    color: theme.palette.text.primary,
    borderLeft: '5px solid transparent',
    transition: theme.transitions.create(['all']),
    '&:hover': {
      textDecoration: 'none',
    }
  },
  activeLink: {
    backgroundColor: theme.palette.type === 'light' ? 'rgba(63, 81, 181, .15)' : 'rgba(51, 51, 51, .15)',
  }
})
});

const menuItems = [
  { icon: <Dashboard />, anchor: 'Главная панель', to: '/' },
  { icon: <ViewList />, anchor: 'Автомобили', to: '/cars' },
  { icon: <Map />, anchor: 'Карта', to: '/maps' },
  { icon: <Report />, anchor: 'Отзывы', to: '/reviews' },
];

interface SideMenuProps {
  isDrawerOpen: boolean;
  closeDrawer: () => void;
}

const SideMenu: React.FC<SideMenuProps> = ({ isDrawerOpen, closeDrawer }: SideMenuProps) => {
  const classes = useStyles();

  return (
    <Drawer
      className={classes.drawer}
      onClose={closeDrawer}
      open={isDrawerOpen}
      classes={{
        paper: classes.drawerPaper,
      }}
    >
      <nav>
        <MenuList>
          {menuItems.map((item) => (
            <MenuItem key={item.to} className={classes.menuItem}>
              <Link
                component={NavLink}
                to={item.to}
                className={classes.link}
                activeClassName={classes.activeLink}
                exact
              >
                <ListItemIcon>{item.icon}</ListItemIcon>
                <Typography variant="inherit">{item.anchor}</Typography>
              </Link>
            </MenuItem>
          ))}
        </MenuList>
      </nav>
    </Drawer>
  );
};

export default SideMenu;
