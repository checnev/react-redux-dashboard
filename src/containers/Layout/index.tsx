import React, { useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import {
  CssBaseline,
  ThemeProvider,
  createMuiTheme,
} from '@material-ui/core';
import { Helmet } from 'react-helmet';

import Header from 'components/Layout/Header';
import SideMenu from 'components/Layout/SideMenu';
import Content from 'components/Layout/Content';
import Footer from 'components/Layout/Footer';

import { toggleDrawer, setThemeMode} from 'actions/layout';
import type { RootState } from 'store';

const useStyles = makeStyles((theme: Theme) => createStyles({
  root: {
    display: 'flex',
    flexDirection: 'column',
    minHeight: '100vh',
  },
}));

interface LayoutProps {
  children: React.ReactNode;
  sidebar?: React.ReactNode;
}

const Layout: React.FC<LayoutProps> = ({ children, sidebar }: LayoutProps) => {
  const classes = useStyles();
  const dispatch = useDispatch();

  const isDrawerOpen = useSelector((state: RootState) => state.layout.isDrawerOpen);
  const themeMode = useSelector((state: RootState) => state.layout.themeMode);

  const openDrawer = useCallback(() => dispatch(toggleDrawer(true)), [dispatch]);
  const closeDrawer = useCallback(() => dispatch(toggleDrawer(false)), [dispatch]);
  const changeTheme = useCallback(() => {
    dispatch(
      setThemeMode(themeMode === 'dark'? 'light' : 'dark')
    );
  }, [dispatch, themeMode]);

  const theme = createMuiTheme({
    palette: {
      type: themeMode,
    },
  });

  return (
    <ThemeProvider theme={theme}>
      <CssBaseline />
      <Helmet htmlAttributes={{ lang: 'ru' }} />
      <div className={classes.root}>
        <Header
          openDrawer={openDrawer}
          changeTheme={changeTheme}
          themeMode={themeMode}
        />
        <SideMenu 
          isDrawerOpen={isDrawerOpen}
          closeDrawer={closeDrawer}
        />
        <Content sidebar={sidebar}>{children}</Content>
        <Footer />
      </div>
    </ThemeProvider>
  );
};

export default Layout;
