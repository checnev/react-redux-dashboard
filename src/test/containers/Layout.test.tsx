import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import pretty from 'pretty';
import Layout from 'containers/Layout';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import state from 'test/data/initialState';

const mockStore = configureMockStore([thunk]);

let container: any = null;
beforeEach(() => {
  container = document.createElement('div');
  document.body.append(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it('LayoutContainer render', () => {
  const store = mockStore(state);
  act(() => {
    render(
      <Provider store={store}>
        <Router>
          <Layout>test content</Layout>
        </Router>
      </Provider>,
      container
    );
  });

  expect(
    pretty(container.innerHTML)
  ).toMatchSnapshot();
});
