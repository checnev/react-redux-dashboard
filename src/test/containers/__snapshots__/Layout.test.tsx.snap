// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutContainer render 1`] = `
"<div class=\\"makeStyles-root-1\\">
  <header class=\\"MuiPaper-root MuiAppBar-root MuiAppBar-positionRelative MuiAppBar-colorPrimary makeStyles-root-2 MuiPaper-elevation4\\">
    <div class=\\"MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters\\"><button class=\\"MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-3 MuiIconButton-colorInherit MuiIconButton-edgeStart\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"menu\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button>
      <h6 class=\\"MuiTypography-root makeStyles-title-4 MuiTypography-h6\\">CarShare</h6><button class=\\"MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"change theme\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69zM12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6c3.31 0 6 2.69 6 6s-2.69 6-6 6z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button>
    </div>
  </header>
  <div class=\\"MuiContainer-root MuiContainer-maxWidthLg\\">
    <div class=\\"MuiGrid-root MuiGrid-container\\">
      <div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12\\">test content</div>
    </div>
  </div>
  <footer class=\\"makeStyles-root-12\\">
    <hr class=\\"MuiDivider-root\\">
    <div class=\\"MuiContainer-root makeStyles-container-13 MuiContainer-maxWidthLg\\">
      <p class=\\"MuiTypography-root MuiTypography-body1 MuiTypography-alignRight\\">CarShare Â© 2021</p>
    </div>
  </footer>
</div>"
`;
